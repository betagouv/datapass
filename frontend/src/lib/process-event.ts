import { getErrorMessages } from '.';
import {
  EnrollmentEvent,
  EventConfiguration,
} from '../config/event-configuration';
import {
  changeEnrollmentState,
  createOrUpdateEnrollment,
  deleteEnrollment,
} from '../services/enrollments';

export const processEvent = async (
  event: EnrollmentEvent,
  eventConfiguration: EventConfiguration,
  enrollment: any,
  updateEnrollment: Function,
  message?: string
): Promise<{
  redirectToHome: boolean;
  successMessages: string[];
  errorMessages: string[];
}> => {
  const resultMessages = {
    errorMessages: [] as string[],
    successMessages: [] as string[],
    redirectToHome: false,
  };

  try {
    let comment = null;
    let enrollmentId = enrollment.id;

    if (eventConfiguration.promptForComment) {
      try {
        comment = message;
      } catch (e) {
        return resultMessages;
      }
    }

    if (eventConfiguration.createOrUpdate) {
      const formattedEnrollment = {
        ...enrollment,
        team_members_attributes: enrollment.team_members,
      };
      const newEnrollment = await createOrUpdateEnrollment({
        enrollment: formattedEnrollment,
      });
      updateEnrollment({ target: { name: 'documents_attributes', value: [] } });
      updateEnrollment(newEnrollment);
      enrollmentId = newEnrollment.id;

      resultMessages.successMessages.push(
        'Votre demande d’habilitation a été sauvegardée.'
      );
    }

    if (eventConfiguration.delete) {
      await deleteEnrollment({ id: enrollmentId });
    }

    if (eventConfiguration.changeEnrollmentState) {
      await changeEnrollmentState({
        event,
        id: enrollmentId,
        comment,
      });
    }

    if (eventConfiguration.redirectToHome) {
      resultMessages.redirectToHome = true;
    }

    if (eventConfiguration.successMessage) {
      resultMessages.successMessages.push(eventConfiguration.successMessage);
    }

    return resultMessages;
  } catch (error) {
    resultMessages.errorMessages.push(...getErrorMessages(error));

    return resultMessages;
  }
};
