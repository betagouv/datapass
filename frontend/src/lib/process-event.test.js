import {
  EnrollmentEvent,
  eventConfigurations,
} from '../config/event-configuration';
jest.mock('../services/enrollments');
// eslint-disable-next-line import/first
import { processEvent } from './process-event';
// eslint-disable-next-line import/first
import {
  changeEnrollmentState,
  deleteEnrollment,
  createOrUpdateEnrollment,
} from '../services/enrollments';

describe('When submitting the enrollment form', () => {
  const enrollment = { id: Symbol(), acl: {} };
  const updateEnrollment = jest.fn();

  beforeEach(() => {
    jest.resetAllMocks();
  });

  describe('with the notify event', () => {
    const event = EnrollmentEvent.notify;
    const eventConfiguration = eventConfigurations.notify;

    it('calls for the enrollment state update', async () => {
      const userMessage = 'La barbe de la femme à Georges Moustaki';

      const output = await processEvent(
        event,
        eventConfiguration,
        enrollment,
        updateEnrollment,
        userMessage,
        false
      );

      expect(changeEnrollmentState).toHaveBeenCalledWith({
        event: 'notify',
        comment: userMessage,
        id: enrollment.id,
      });
      expect(output).toMatchSnapshot();
    });
  });

  describe('with the destroy event', () => {
    const event = EnrollmentEvent.destroy;
    const eventConfiguration = eventConfigurations.destroy;

    it('calls the delete endpoint', async () => {
      const output = await processEvent(
        event,
        eventConfiguration,
        enrollment,
        updateEnrollment
      );

      expect(deleteEnrollment).toHaveBeenCalledWith({
        id: enrollment.id,
      });
      expect(output).toMatchSnapshot();
    });
  });

  describe('with the update event', () => {
    const event = EnrollmentEvent.update;
    const eventConfiguration = eventConfigurations.update;

    const enrollmentToUpdate = { ...enrollment, acl: { update: true } };

    it('calls the update endpoint', async () => {
      createOrUpdateEnrollment.mockResolvedValue(enrollmentToUpdate);

      const output = await processEvent(
        event,
        eventConfiguration,
        enrollmentToUpdate,
        updateEnrollment
      );

      expect(createOrUpdateEnrollment).toHaveBeenCalledWith({
        enrollment: enrollmentToUpdate,
      });
      expect(output).toMatchSnapshot();
    });

    it('displays an error if update fails', async () => {
      createOrUpdateEnrollment.mockRejectedValue("Pas d'update désolé");

      const output = await processEvent(
        event,
        eventConfiguration,
        enrollmentToUpdate,
        updateEnrollment
      );

      expect(output).toMatchSnapshot();
    });
  });
});
