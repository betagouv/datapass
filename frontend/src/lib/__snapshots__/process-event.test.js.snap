// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When submitting the enrollment form with the destroy event calls the delete endpoint 1`] = `
Object {
  "errorMessages": Array [],
  "redirectToHome": true,
  "successMessages": Array [],
}
`;

exports[`When submitting the enrollment form with the notify event calls for the enrollment state update 1`] = `
Object {
  "errorMessages": Array [],
  "redirectToHome": true,
  "successMessages": Array [
    "Votre message a été envoyé. Vous recevrez une notification dans votre boite mail lorsqu’il sera traité.",
  ],
}
`;

exports[`When submitting the enrollment form with the update event calls the update endpoint 1`] = `
Object {
  "errorMessages": Array [],
  "redirectToHome": false,
  "successMessages": Array [
    "Votre demande d’habilitation a été sauvegardée.",
  ],
}
`;

exports[`When submitting the enrollment form with the update event displays an error if update fails 1`] = `
Object {
  "errorMessages": Array [
    "Une erreur inconnue est survenue. Merci de réessayer ultérieurement. Vous pouvez également nous signaler cette erreur par mail à contact@api.gouv.fr.",
  ],
  "redirectToHome": false,
  "successMessages": Array [],
}
`;
